Source/OCMock/OCPartialMockObject.m

```diff
@@ -29,12 +29,20 @@ @implementation OCPartialMockObject
 
 - (id)initWithObject:(NSObject *)anObject
 {
-    NSParameterAssert(anObject != nil);
-    Class const class = [self classToSubclassForObject:anObject];
-    [self assertClassIsSupported:class];
-	[super initWithClass:class];
-	realObject = [anObject retain];
-    [self prepareObjectForInstanceMethodMocking];
+    @try
+    {
+        NSParameterAssert(anObject != nil);
+        Class const class = [self classToSubclassForObject:anObject];
+        [self assertClassIsSupported:class];
+        [super initWithClass:class];
+        realObject = [anObject retain];
+        [self prepareObjectForInstanceMethodMocking];
+    }
+    @catch(NSException *e)
+    {
+        [OCMockObject removeAMockToStop:self];
+        [e raise];
+    }
```

Comment:
I agree that it's inelegant, but the only reason that is there is because of the tests for OCMock itself that test exceptions thrown in Mock objects. The standard Cocoa convention is that exceptions signal programmer error and are not intended to be recovered from, and I would argue that goes doubly for throwing from an init method.  By default in Objective C, ARC is not exception-safe; I know this is not the case for Objective C++, but I willing to wager that the amount of folks who are throwing in init from an OCMockObject subclass and expect tests to continue cleanly is vanishingly small (present company excluded of course :) ).

FWIW I'm looking at an extremely large codebase with several hundred iOS/macOS targets with over 45K reference to OCMock in on form or another we have only one subclass of an OCMockObject in any form that I can find (and it would not be affected by this change).